crd/cluster.go

```diff
@@ -0,0 +1,76 @@
+package crd
+
+import (
+	"fmt"
+	"os"
+	"time"
+
+	apiextensions "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	apiextensionsclient "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/apimachinery/pkg/util/yaml"
+)
+
+func YAMLToCRD(filepath string) (*apiextensions.CustomResourceDefinition, error) {
+	f, err := os.Open(filepath)
+	if err != nil {
+		return nil, err
+	}
+
+	crd := &apiextensions.CustomResourceDefinition{}
+	yaml.NewYAMLOrJSONDecoder(f, 32).Decode(crd)
+
+	return crd, nil
+}
+
+func MySQLClusterCreateCRD(clientset *apiextensionsclient.Clientset) error {
+	crd, err := YAMLToCRD("artifacts/mysqlcrd.yaml")
+	if err != nil {
+		return err
+	}
+
+	_, err = clientset.ApiextensionsV1beta1().CustomResourceDefinitions().Create(crd)
+	if err != nil {
+		if apierrors.IsAlreadyExists(err) {
+			return nil
+		}
+		return err
+	}
+
+	err = waitForCRDEstablished(clientset, crd.ObjectMeta.Name)
+	if err != nil {
+		deleteErr := clientset.ApiextensionsV1beta1().CustomResourceDefinitions().Delete(crd.ObjectMeta.Name, nil)
+		if deleteErr != nil {
+			return errors.NewAggregate([]error{err, deleteErr})
+		}
+		return err
+	}
+
+	return nil
+}
+
+func waitForCRDEstablished(clientset *apiextensionsclient.Clientset, CRDName string) error {
```

Comment:
It might be a good function to implement on clientset itself
```go
func (clientset *apiextensionsclient.Clientset) waitForCRDEstablished(CRDname string) error {
    ...
}
```